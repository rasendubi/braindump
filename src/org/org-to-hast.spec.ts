import YAML from 'yaml';
import { Type } from 'yaml/util';

import unified from 'unified';
import orgParse from './unified-org-parse';
import org2rehype from './unified-org-rehype';
import raw from 'rehype-raw';
import format from 'rehype-format';
import html from 'rehype-stringify';

const processor = unified()
  .use(orgParse)
  .use(org2rehype)
  .use(raw)
  .use(format)
  .use(html);

YAML.scalarOptions.str.defaultType = Type.QUOTE_DOUBLE;
YAML.scalarOptions.str.defaultKeyType = Type.PLAIN;

expect.addSnapshotSerializer({
  test(value) {
    return typeof value === 'string';
  },
  print(value) {
    return value as string;
  },
});

const hastTest = (name: string, input: string) => {
  it(name, () => {
    const result = processor.processSync(input).contents;
    expect(result).toMatchSnapshot();
  });
};
hastTest.skip = (name: string, input: string) => {
  it.skip(name, () => {
    const result = processor.processSync(input).contents;
    expect(result).toMatchSnapshot();
  });
};
hastTest.only = (name: string, input: string) => {
  it.only(name, () => {
    const result = processor.processSync(input).contents;
    expect(result).toMatchSnapshot();
  });
};
hastTest.todo = (name: string, _input?: string) => {
  it.todo(name);
};

describe('org/org-to-hast', () => {
  hastTest('empty', ``);

  hastTest('paragraph', `hello`);

  hastTest('headline', `* hi`);

  hastTest(
    'multiple headlines',
    `* hi
** there
* hello
*** world
`
  );

  hastTest(
    'complex headline',
    `* TODO [#A] headline /italic/ title :some:tags:`
  );

  hastTest(
    'headline with sections',
    `hello
* hi
section
** hello
another section`
  );

  hastTest(
    'planning',
    `* headline
CLOSED: [2019-03-13 Wed 23:48] SCHEDULED: [2019-03-13 Wed] DEADLINE: [2019-03-14 Thu]`
  );

  hastTest(
    'property drawer',
    `* headline
:PROPERTIES:
:CREATED: [2019-03-13 Wed 23:57]
:END:`
  );

  hastTest(
    'custom drawer',
    `:MYDRAWER:
hello /there/
:END:`
  );

  hastTest('list', `- hello`);

  hastTest('ordered list', `1. one`);

  hastTest(
    'nested list',
    `
- hello
  - world
  - blah
- hi
`
  );

  hastTest('link', `https://example.com`);

  hastTest('link mixed with text', `hello http://example.com blah`);

  hastTest(
    'src block',
    `#+begin_src c
,*a = b;
printf("%d\\n", *a);
#+end_src`
  );

  hastTest(
    'verse block',
    `#+BEGIN_VERSE
 Great clouds overhead
 Tiny black birds rise and fall
 Snow covers Emacs

    ---AlexSchroeder
#+END_VERSE`
  );

  hastTest(
    'center block',
    `#+begin_center
hello
#+end_center`
  );

  hastTest(
    'comment block',
    `#+begin_comment
hello
#+end_comment`
  );

  hastTest(
    'remove common src block offset',
    `
  #+begin_src
    hello
      world
  #+end_src`
  );

  hastTest(
    'example block',
    `#+begin_example
example
#+end_example`
  );

  hastTest(
    'export block',
    `#+begin_export
export
#+end_export`
  );

  hastTest(
    'export html block',
    `
#+begin_export html
<abbr title="World Health Organization">WHO</abbr> was founded in 1948.
#+end_export
`
  );

  hastTest(
    'special block',
    `#+begin_blah
hello
#+end_blah`
  );

  hastTest(
    'blockquote',
    `#+begin_quote
hello, world!
#+end_quote`
  );

  hastTest('keywords', `#+TITLE: blah`);

  hastTest(
    'emphasis',
    `/Consider/ ~t*h*e~ *following* =example= +strike+ _under_`
  );

  hastTest('images', `[[./image.png]]`);

  describe('timestamps', () => {
    hastTest('inactive', `[2021-01-07 Thu]`);
    hastTest('inactive-range', `[2021-01-07 Thu]--[2021-01-08 Fri]`);
    hastTest('active', `<2021-01-07 Thu>`);
    hastTest('active-range', `<2021-01-07 Thu>--<2021-01-09 Sat>`);
    hastTest('with time', `[2021-01-07 Thu 19:36]`);
    hastTest('time range', `[2021-01-07 Thu 19:36-20:38]`);
    hastTest.todo('diary');
  });

  hastTest(
    'table',
    `
| head1  | head2 |
|--------+-------|
| value1 | value2 |
`
  );

  hastTest(
    'comments',
    `hello
# comment
there`
  );

  hastTest('fixed-width', `: hello`);

  hastTest(
    'clock',
    `CLOCK: [2020-12-22 Tue 09:07]--[2020-12-22 Tue 11:10] =>  2:03`
  );

  hastTest(
    `latex environment`,
    `\\begin{hello}
some text
\\end{hello}`
  );

  hastTest('horizontal rule', `-----`);

  hastTest(
    'diary sexp',
    `%%(diary-anniversary 10 31 1948) Arthur's birthday (%d years old)`
  );

  hastTest(
    'latex-fragment',
    `If $a^2=b$ and \\( b=2 \\), then the solution must be
either $$ a=+\\sqrt{2} $$ or \\[ a=-\\sqrt{2} \\].`
  );

  hastTest('entity', `\\Agrave`);
});
