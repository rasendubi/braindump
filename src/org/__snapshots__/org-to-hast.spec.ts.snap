// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`org/org-to-hast blockquote 1`] = `
<div>
<blockquote>
  <p>hello, world!</p>
</blockquote>
</div>
`;

exports[`org/org-to-hast center block 1`] = `
<div>
<div class="center">
  <p>hello</p>
</div>
</div>
`;

exports[`org/org-to-hast clock 1`] = `<div></div>`;

exports[`org/org-to-hast comment block 1`] = `<div></div>`;

exports[`org/org-to-hast comments 1`] = `
<div>
<p>hello</p>
<p>there</p>
</div>
`;

exports[`org/org-to-hast complex headline 1`] = `
<div>
<h1><span class="todo-keyword TODO">TODO</span> <span class="priority">[A]</span> headline <em>italic</em> title   <span class="tags"><span class="tag tag-some">some</span> <span class="tag tag-tags">tags</span></span></h1>
</div>
`;

exports[`org/org-to-hast custom drawer 1`] = `<div></div>`;

exports[`org/org-to-hast diary sexp 1`] = `<div></div>`;

exports[`org/org-to-hast emphasis 1`] = `
<div>
<p><em>Consider</em> <code class="inline-code">t*h*e</code> <strong>following</strong> <code class="inline-verbatim">example</code> <del>strike</del> <span style="text-decoration: underline;">under</span></p>
</div>
`;

exports[`org/org-to-hast empty 1`] = `<div></div>`;

exports[`org/org-to-hast example block 1`] = `
<div>
<div class="exampe">example</div>
</div>
`;

exports[`org/org-to-hast export block 1`] = `<div></div>`;

exports[`org/org-to-hast export html block 1`] = `<div><abbr title="World Health Organization">WHO</abbr> was founded in 1948.</div>`;

exports[`org/org-to-hast fixed-width 1`] = `
<div>
<pre>hello</pre>
</div>
`;

exports[`org/org-to-hast headline 1`] = `
<div>
<h1>hi</h1>
</div>
`;

exports[`org/org-to-hast headline with sections 1`] = `
<div>
<p>hello</p>
<h1>hi</h1>
<p>section</p>
<h2>hello</h2>
<p>another section</p>
</div>
`;

exports[`org/org-to-hast horizontal rule 1`] = `
<div>
<hr>
</div>
`;

exports[`org/org-to-hast images 1`] = `
<div>
<p>
  <img src="./image.png">
</p>
</div>
`;

exports[`org/org-to-hast keywords 1`] = `<div></div>`;

exports[`org/org-to-hast latex environment 1`] = `
<div>
<div class="math math-display">
  \\begin{hello}
  some text
  \\end{hello}
</div>
</div>
`;

exports[`org/org-to-hast latex-fragment 1`] = `
<div>
<p>
  If <span class="math math-inline">$a^2=b$</span> and <span class="math math-inline">\\( b=2 \\)</span>, then the solution must be
  either <span class="math math-inline">$$ a=+\\sqrt{2} $$</span> or <span class="math math-inline">\\[ a=-\\sqrt{2} \\]</span>.
</p>
</div>
`;

exports[`org/org-to-hast link 1`] = `
<div>
<p><a href="https://example.com">https://example.com</a></p>
</div>
`;

exports[`org/org-to-hast link mixed with text 1`] = `
<div>
<p>hello <a href="http://example.com">http://example.com</a> blah</p>
</div>
`;

exports[`org/org-to-hast list 1`] = `
<div>
<ul>
  <li>
    <p>hello</p>
  </li>
</ul>
</div>
`;

exports[`org/org-to-hast multiple headlines 1`] = `
<div>
<h1>hi</h1>
<h2>there</h2>
<h1>hello</h1>
<h3>world</h3>
</div>
`;

exports[`org/org-to-hast nested list 1`] = `
<div>
<ul>
  <li>
    <p>hello</p>
    <ul>
      <li>
        <p>world</p>
      </li>
      <li>
        <p>blah</p>
      </li>
    </ul>
  </li>
  <li>
    <p>hi</p>
  </li>
</ul>
</div>
`;

exports[`org/org-to-hast ordered list 1`] = `
<div>
<ol>
  <li>
    <p>one</p>
  </li>
</ol>
</div>
`;

exports[`org/org-to-hast paragraph 1`] = `
<div>
<p>hello</p>
</div>
`;

exports[`org/org-to-hast planning 1`] = `
<div>
<h1>headline</h1>
</div>
`;

exports[`org/org-to-hast property drawer 1`] = `
<div>
<h1>headline</h1>
</div>
`;

exports[`org/org-to-hast remove common src block offset 1`] = `
<div>
<pre><code>hello
  world
</code></pre>
</div>
`;

exports[`org/org-to-hast special block 1`] = `
<div>
<div>
  <p>hello</p>
</div>
</div>
`;

exports[`org/org-to-hast src block 1`] = `
<div>
<pre><code class="language-c">a = b;
printf("%d\\n", *a);
</code></pre>
</div>
`;

exports[`org/org-to-hast table 1`] = `
<div>
<table>
  <thead>
    <tr>
      <th>head1</th>
      <th>head2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>value1</td>
      <td>value2</td>
    </tr>
  </tbody>
</table>
</div>
`;

exports[`org/org-to-hast timestamps active 1`] = `
<div>
<p><span class="timestamp">&#x3C;2021-01-07 Thu></span></p>
</div>
`;

exports[`org/org-to-hast timestamps active-range 1`] = `
<div>
<p><span class="timestamp">&#x3C;2021-01-07 Thu>--&#x3C;2021-01-09 Sat></span></p>
</div>
`;

exports[`org/org-to-hast timestamps inactive 1`] = `
<div>
<p><span class="timestamp">[2021-01-07 Thu]</span></p>
</div>
`;

exports[`org/org-to-hast timestamps inactive-range 1`] = `
<div>
<p><span class="timestamp">[2021-01-07 Thu]--[2021-01-08 Fri]</span></p>
</div>
`;

exports[`org/org-to-hast timestamps time range 1`] = `
<div>
<p><span class="timestamp">[2021-01-07 Thu 19:36-20:38]</span></p>
</div>
`;

exports[`org/org-to-hast timestamps with time 1`] = `
<div>
<p><span class="timestamp">[2021-01-07 Thu 19:36]</span></p>
</div>
`;

exports[`org/org-to-hast verse block 1`] = `
<div>
<p class="verse">
  Great clouds overhead
  Tiny black birds rise and fall
  Snow covers Emacs
  ---AlexSchroeder
</p>
</div>
`;
